### PR TITLE
feat: ledger

### PR DESCRIPTION
**Pull request by jdutchak**
As part of https://github.com/masa-finance/masa-oracle/issues/345 we have decided to go with our on-chain ledger rather then using another third party ledger.

The key takeaways are:

our own ledger that can store a limited set of data in the on-chain ledger. The ledger is implemented similarly to what EdgeVPN uses.
The ledger can be used to store a CID (content addressable ID) to reference a data that is carried over with the ledger. that for instance can be used to store big files or content that doesn't fit the ledger (see also https://github.com/masa-finance/masa-oracle/issues/345#issuecomment-2167399111 )
The ledger can be written only by trusted nodes, but can be read by all nodes of the network
For development, we can also point to an IPFS CID that is then pinned by the oracle nodes. That would allow us to quickly iterate in the short term while we develop a CID system inside our own protocol.
This card is about wrapping up the esperimental PR work in https://github.com/masa-finance/masa-oracle/pull/350 and have it merged in the repository.

Describe alternatives you've considered
Using an external blockchain as a ledger (see discussions in https://github.com/masa-finance/masa-oracle/issues/345)

Acceptance criteria
 https://github.com/masa-finance/masa-oracle/pull/350 is wrapped up and merged in the codebase

### CODE REVIEW

**Comment by mudler** [comment]
@jdutchak can you add a description to the PR and have the title reflect what's the changeset? thank you!

File: cmd/masa-node/main.go [hunk 1]
```diff
@@ -89,7 +89,7 @@ func main() {
 	// Init cache resolver
 	go db.InitResolverCache(node, keyManager)
 	// Subscribe to blocks -- moving to new ticket
```
Comment (mudler) [inline cmd/masa-node/main.go:91]: maybe we can drop the comment too?
```suggestion
	// Subscribe to blocks
```

File: pkg/chain/block.go [hunk 2]
```diff
@@ -13,7 +13,7 @@ type Block struct {
 	Data  []byte //	this block's data
 	Hash  []byte //	this block's hash
 	Link  []byte //	the hash of the last block in the chain. this is the key part that links the blocks together
-	Nonce int64  //	the nonce used to sing the block. useful for verification
+	Nonce int64  //	the nonce used to sing the block for verification
```
Comment (mudler) [inline pkg/chain/block.go:16]:
```suggestion
	Nonce int64  //	the nonce used to sign the block for verification
```

File: .github/workflows/e2e.yaml [hunk 3]
```diff
@@ -0,0 +1,28 @@
+name: Go
+
+on:
+  push:
```
Comment (mudler) [inline .github/workflows/e2e.yaml:4]: Shall we run it to all the branches? I'd think would be quite useful to know if the `main` branch is in good shape too.

Comment (jdutchak) [reply .github/workflows/e2e.yaml:4]: i believe we only have test and PR branches - do you think we should always run tests ?

Comment (jdutchak) [reply .github/workflows/e2e.yaml:4]: if so we can do this 
```
on:
  push:
    branches:
      - '**'
  pull_request:
    branches:
      - '**'
```

Comment (mudler) [reply .github/workflows/e2e.yaml:4]: > i believe we only have test and PR branches - do you think we should always run tests ?

I'm a bit on the yes side, but maybe @masa-finance/protocol have different thoughts in here?

File: .github/workflows/e2e.yaml [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    - name: Set up Go
+      uses: actions/setup-go@v2
+      with:
+        go-version: '^1.22'
+
+    - name: Install dependencies
+      run: go mod tidy
+
+    - name: Run tests
+      run: go test ./...
```
Comment (mudler) [inline .github/workflows/e2e.yaml:27]: Great addition!

just re-reading this now, but this doesn't look like e2e tests isn't it? This is actually covering #369  . Please rename the pipeline file accordingly, would be even better to keep in a separate PR as doesn't seems we are bringing new tests in this PR that are covered by it?

```

Run go test ./...
?   	github.com/masa-finance/masa-oracle/cmd/masa-node	[no test files]
?   	github.com/masa-finance/masa-oracle/cmd/masa-node-cli	[no test files]
?   	github.com/masa-finance/masa-oracle/cmd/tagger	[no test files]
?   	github.com/masa-finance/masa-oracle/docs	[no test files]
?   	github.com/masa-finance/masa-oracle/pkg/api	[no test files]
?   	github.com/masa-finance/masa-oracle/pkg/chain	[no test files]
?   	github.com/masa-finance/masa-oracle/pkg/config	[no test files]
ok  	github.com/masa-finance/masa-oracle/pkg	0.026s
?   	github.com/masa-finance/masa-oracle/pkg/db	[no test files]
?   	github.com/masa-finance/masa-oracle/pkg/llmbridge	[no test files]
ok  	github.com/masa-finance/masa-oracle/pkg/consensus	0.015s
?   	github.com/masa-finance/masa-oracle/pkg/network	[no test files]
?   	github.com/masa-finance/masa-oracle/pkg/pubsub	[no test files]
?   	github.com/masa-finance/masa-oracle/pkg/scrapers/discord	[no test files]
?   	github.com/masa-finance/masa-oracle/pkg/scrapers/reddit	[no test files]
?   	github.com/masa-finance/masa-oracle/pkg/scrapers/twitter	[no test files]
?   	github.com/masa-finance/masa-oracle/pkg/scrapers/web	[no test files]
?   	github.com/masa-finance/masa-oracle/pkg/staking	[no test files]
ok  	github.com/masa-finance/masa-oracle/pkg/masacrypto	0.016s
?   	github.com/masa-finance/masa-oracle/pkg/workers	[no test files]
?   	github.com/masa-finance/masa-oracle/pkg/workers/messages	[no test files]
?   	github.com/masa-finance/masa-oracle/tests	[no test files]
?   	github.com/masa-finance/masa-oracle/tools	[no test files]
ok  	github.com/masa-finance/masa-oracle/pkg/tests	1.306s
```

File: pkg/network/discover.go [hunk 5]
```diff
@@ -97,8 +97,9 @@ func Discover(ctx context.Context, host host.Host, dht *dht.IpfsDHT, protocol pr
 					Addrs: availPeer.Addrs,
 				}
 				hostAddrInfo := peer.AddrInfo{
-					ID:    host.ID(),
-					Addrs: []multiaddr.Multiaddr{},
+					ID: host.ID(),
+					//Addrs: []multiaddr.Multiaddr{},
```
Comment (mudler) [inline pkg/network/discover.go:101]: commented code added

File: pkg/pubsub/worker_tracker.go [hunk 6]
```diff
@@ -26,7 +26,7 @@ type WorkerEventTracker struct {
 
 // HandleMessage implements subscription WorkerEventTracker handler
 func (h *WorkerEventTracker) HandleMessage(m *pubsub.Message) {
-	logrus.Infof("workerStream -> Received work from: %s", m.ReceivedFrom)
+	// logrus.Infof("workerStream -> Received work from: %s", m.ReceivedFrom)
```
Comment (mudler) [inline pkg/pubsub/worker_tracker.go:29]: Move to debug?

File: pkg/tests/scraped_tweets.json [hunk 7]
```diff
@@ -0,0 +1 @@
...de $MASA tokens to Base ETH which can easily be sent to any CEX of your choice.\u003cbr\u003e\u003cbr\u003eHere, I will take it bit by bit to explain the process via \u003ca href=\"https://twitter.com/Uniswap\"\u003e@Uniswap\u003c/a\u003e in this 🧵 ....let's go 😎","ID":"1796651254048760256","InReplyToStatus":null,"InReplyToStatusID":"","IsQuoted":false,"IsPin":false,"IsReply":false,"IsRetweet":false,"IsSelfThread":false,"Likes":0,"Name":"FRANKSTAINZ.SOL 🕋 🪐🇫🇷🪐","Mentions":[{"ID":"984188226826010624","Username":"Uniswap","Name":"Uniswap Labs 🦄"}],"PermanentURL":"https://twitter.com/marcelokpe/status/1796651254048760256","Photos":null,"Place":null,"QuotedStatus":null,"QuotedStatusID":"","Replies":0,"Retweets":0,"RetweetedStatus":null,"RetweetedStatusID":"","Text":"Lately I've had some friends asking questions on how to trade $MASA tokens to Base ETH which can easily be sent to any CEX of your choice.\n\nHere, I will take it bit by bit to explain the process via @Uniswap in this 🧵 ....let's go 😎","Thread":null,"TimeParsed":"2024-05-31T21:13:33Z","Timestamp":1717190013,"URLs":null,"UserID":"1448419864289366028","Username":"marcelokpe","Videos":null,"Views":48,"SensitiveContent":false}]
```
Comment (mudler) [inline pkg/tests/scraped_tweets.json:1]: do we need so many tweets for test fixtures?

**Review by mudler — CHANGES_REQUESTED** [review]
Changes overall looks good, however:

- Above nitpicks
- Please split the work of #369 in a separate PR
- The title and description are copied verbatim from the issue #380 . The PR should have a title/description that shows the changeset introduced. Please see https://www.conventionalcommits.org/en/v1.0.0/ for guidance

File: RELEASE_NOTES.md [hunk 8]
```diff
@@ -2,26 +2,26 @@
 
 [All Releases](https://github.com/masa-finance/masa-oracle/releases)
 
-## [0.0.9-beta](https://github.com/masa-finance/masa-oracle/releases) (2024)
+## [0.5.0](https://github.com/masa-finance/masa-oracle/releases) (2024)
```
Comment (mudler) [inline RELEASE_NOTES.md:5]: Can we please avoid to mix release notes changes among PRs?  Let's keep the scope of the PR tight to what we are adding/changing. 

This belongs to another PR or a follow-up to update the release notes.

File: cmd/masa-node-cli/components.go [hunk 9]
```diff
@@ -122,7 +122,7 @@ const logo = `
 const (
 	subtitle   = `masa oracle client`
 	navigation = `[yellow]use keys or mouse to navigate`
-	mouse      = `[green]v0.0.9-beta`
+	mouse      = `[green]v0.5.0`
```
Comment (mudler) [inline cmd/masa-node-cli/components.go:125]: ditto here, we shouldn't bump the version here

**Review by restevens402 — APPROVED** [review]
Looks good. Blocks are clean and straightforward. Let's go!

File: cmd/masa-node-cli/components.go (hunk 9 above)
Comment (restevens402) [reply cmd/masa-node-cli/components.go:125]: this was updated to reflect that version

File: RELEASE_NOTES.md (hunk 8 above)
Comment (restevens402) [reply RELEASE_NOTES.md:5]: Makes sense. Will do this going forward.
The version was updated accordingly as specified above.

File: cmd/masa-node-cli/components.go (hunk 9 above)
Comment (mudler) [reply cmd/masa-node-cli/components.go:125]: I get it, but It really should be a separate PR :)

File: .github/workflows/e2e.yaml [hunk 10]
```diff
@@ -0,0 +1,27 @@
+name: Go
```
Comment (mudler) [inline .github/workflows/e2e.yaml:1]: File is named `e2e`, but this is running unit tests

File: pkg/api/handlers_node.go [hunk 11]
```diff
@@ -443,3 +443,30 @@ func (api *API) GetNodeApiKey() gin.HandlerFunc {
 		})
 	}
 }
+
+// wip
+// peerInfos := node.DHT.RoutingTable().GetPeerInfos()
```
Comment (mudler) [inline pkg/api/handlers_node.go:448]: do we need to keep this code around?

File: pkg/workers/workers.go [hunk 12]
```diff
@@ -188,12 +188,19 @@ func updateRecords(node *masa.OracleNode, workEvent db.WorkEvent) {
...
+		}
+
 	}
 
 	var nodeData pubsub.NodeData
 	nodeDataBytes, err := db.GetCache(context.Background(), workEvent.PeerId)
 	if err != nil {
-		nodeData = *node.NodeTracker.GetNodeData(workEvent.PeerId)
+		if workEvent.PeerId != "" {
+			nodeData = *node.NodeTracker.GetNodeData(workEvent.PeerId)
```
Comment (mudler) [inline pkg/workers/workers.go:202]: This dereference is making assumption that we always return a non-nil pointer. While that might be correct for 99% of the cases, it's better to be defensive here and check it what `GetNodeData` is returning is not nil

Comment (mudler) [inline pkg/workers/workers.go:202]:
```suggestion
			nodeDataPtr = node.NodeTracker.GetNodeData(workEvent.PeerId)
			if nodeDataPtr != nil {
			   nodeData = *nodeDataPtr
			 }
```

**Review by mudler — CHANGES_REQUESTED** [review]
Few changes requested but ignored - some of them are nitpicks, but I think some of these needs to be addressed before merging:

- please split the release the release note changes to a separate PR
- update the PR title/description to match the changeset
- brittle pointer dereference

Otherwise looks good - nice job!

**Comment by mudler** [comment]
re: commitlint, if you rebase on top of `test` now that should be gone, the CI checks now PR title/description rather then single commits

File: pkg/api/handlers_node.go (hunk 11 above)
Comment (jdutchak) [reply pkg/api/handlers_node.go:460]: removed and saved elsewhere for wip

File: pkg/workers/workers.go (hunk 12 above)
Comment (jdutchak) [reply pkg/workers/workers.go]: resolved by B

**Review by jdutchak — COMMENTED** [review]
done

**Review by restevens402 — DISMISSED** [review]
Everything looks like it is resolved.
I looked into what it would take to move all the version updates into another branch and it would take some time. I understand we want to do this going forward, can we keep in this PR and we will make sure to do that separately in future releases.

File: pkg/workers/workers.go [hunk 13]
```diff
@@ -193,17 +199,18 @@ func updateRecords(node *masa.OracleNode, workEvent db.WorkEvent) {
...
-	}
+	exists, _ := db.ReadData(node, workEvent.CID)
+	// we don't need to check for err since !exists gives an err also - we only need to know if the record exists or not in this context
 	if exists == nil {
 		err := db.WriteData(node, workEvent.CID, workEvent.Payload)
 		if err != nil {
 			logrus.Errorf("Failed to write data for CID %s: %v", workEvent.CID, err)
 			return
 		}
+		err = node.PubSubManager.Publish(config.TopicWithVersion(config.BlockTopic), workEvent.Payload)
```
Comment (mudler) [inline pkg/workers/workers.go:210]: @jdutchak can you confirm this bit? It looks good here, but just wondering how that bubbled up by rebasing

**Review by mudler — APPROVED** [review]
Looking good! great job :partying_face:

**Review by mudler — APPROVED** [review]

**Review by mudler — APPROVED** [review]
